Add support for instagram.com user profiles and pages

iamleot opened the pull request:
The extractor scrapes `instagram.com/<user>' timelines and
`instagram.com/p/<shortcode>' by mimicking the behaviour of a web
browser and extracting the sharedData JSON of the single pages.

Please note that this mean that for user timelines we also do an
extra request to the `instagram.com/p/<shortcode>' page but this
permit to have consistent (and all) information about the media
fetched.

The MD5 logic used for X-Instagram-GIS was documented in

 <https://stackoverflow.com/questions/49786980/>

The current known (and documented via TODO comment) limitation of
the extractor is that it treats `GraphSidecar' media (that contains
multiple images) like `GraphImage'.  So, for `GraphSidecar' media - at the
moment -only an image is retrieved, not all of them.

(I would prefer to add support for it once this patch is reviewed
and imported!)


Thank you very much!

iamleot commented:
Unfortunately it seems that the first two tests of InstagramPostpageExtractor fails.
I guess that the URLs are not stable.  Any suggestion on how to address that?
(I think that we can check their ID and shortcodes instead).

iamleot commented:
I have adjusted the tests in order to avoid checking the URL for `GraphImage` and `GraphSidecar`.  While here I have also added checks for keywords in `GraphVideo` too (in that case the URL is stable because it is a `ytdl:` one with just the shortcode).

I have added a second commit to also check the count of `InstagramProfilepageExtractor`.

mikf commented on gallery_dl/extractor/instagram.py:
```diff
@@ -0,0 +1,189 @@
+# it under the terms of the GNU General Public License version 2 as
+# published by the Free Software Foundation.
+
+"""Extract images from Instagram"""
```
The module docstring should end in the full root URL of the site in question.
`Instagram` -> `https://www.instagram.com/`

(This is used to auto-generate `docs/supportedsites.rst`. You should run `./scripts/build_supportedsites.py` and see how this turns out.)

iamleot replied:
I have adjusted them respectively to `image` and and `user` (so
the description in `docs/supportedsites.rst` will also looks nicer
based on the known categories).
I have also renamed the `Instagram*Extractor`s accordingly
(`InstagramPostpageExtractor` -> `InstagramImageExtractor`,
`InstagramProfilepageExtractor` -> `InstagramUserExtractor`). This was
pointed out by `test/test_extractor.py`.

mikf commented on gallery_dl/extractor/instagram.py:
```diff
@@ -0,0 +1,189 @@
+    root = "https://www.instagram.com"
+
+    def __init__(self, match):
+        Extractor.__init__(self)
```
This appears rather pointless. Don't use an `__init__` function if there is nothing to do.

iamleot replied:
I have removed it and adjusted the other `__init__()` to call
`InstagramExtractor.__init__(self)` instead of
`InstagramExtractor.__init__(self, match)`.

mikf commented on gallery_dl/extractor/instagram.py:
```diff
@@ -0,0 +1,189 @@
+            'height': int(media['dimensions']['height']),
+            'width': int(media['dimensions']['width']),
+            'comments': int(media['edge_media_to_comment']['count']),
+            'likes': int(media['edge_media_preview_like']['count']),
```
Use `text.parse_int()` to convert numbers to integer. It doesn't throw an exception if things go wrong.

iamleot replied:
I have replaced all `int()` to `text.parse_int()`.

mikf commented on gallery_dl/extractor/instagram.py:
```diff
@@ -0,0 +1,189 @@
+                '66eb9403e44cc12e5b5ecda48b667d41',
+                variables,
+            )
+            shared_data = json.loads(self.request(url, headers=headers).text)
```
The `response` object returned by `self.request()` has a `.json()` method. Use that instead of "manually" loading the JSON result. -> `shared_data = self.request(...).json()`

iamleot replied:
Whoops, I have adjusted it as suggested!

iamleot replied:
I have converted it to a tuple.

mikf commented on gallery_dl/extractor/instagram.py:
```diff
@@ -0,0 +1,189 @@
+
+    def instagrams(self):
+        url = '{}/p/{}/'.format(self.root, self.shortcode)
+        return [self._extract_postpage(url)]
```
Not too important, but you can use a tuple instead of a list. (It's a tiny bit more lightweight)
`[...]` -> `(...,)`

iamleot replied:
I have converted it to a tuple.

mikf commented on gallery_dl/extractor/instagram.py:
```diff
@@ -0,0 +1,189 @@
+class InstagramProfilepageExtractor(InstagramExtractor):
+    """Extractor for ProfilePage"""
+    subcategory = "profilepage"
+    pattern = [r"(?:https?://)?(?:www\.)?instagram\.com/([^/]+)/?$"]
```
Use a negative lookahead if you don't want to match single `p/`s here.
`instagram\.com/(?!p/)([^/?&#]+)`

iamleot replied:
TIL about look{ahead,behind}, nice catch, I have adjusted it! :)

iamleot commented:
Hello Mike,

Mike Fährmann writes:
> mikf commented on this pull request.
>
> Thanks for the instagram support! It mostly it looks really solid, but there some small things that should be changed.
>

Thank you very much for your review!  Sure, let's adjust it!

> For example extractor- and subcategory names: the current ones (`postpage`, `profilepage`) are a bit too verbose, I think. Just `post` and `profile` are better, but I would even consider calling them `image` and `user` to be in line with other extractors. Your choice.
>

I have adjusted them respectively to `image` and and `user` (so
the description in `docs/supportedsites.rst` will also looks nicer
based on the known categories).
I have also renamed the `Instagram*Extractor`s accordingly
(`InstagramPostpageExtractor` -> `InstagramImageExtractor`,
`InstagramProfilepageExtractor` -> `InstagramUserExtractor`). This was
pointed out by `test/test_extractor.py`.

> >  I guess that the URLs are not stable. Any suggestion on how to address that?
>
> Use `pattern` instead of `url` and specify a regex that should match all returned URLs. ([example](https://github.com/mikf/gallery-dl/blob/e80ee77d71d55f9e186927a5df43206c1b24e016/gallery_dl/extractor/artstation.py#L126-L130))
>

Great! I've added a `pattern:` in all tests that omitted it (I have
only tested it via Tor and it seems that only the subdomain part
seems to differ (e.g. I had `scontent-mxp1-1.` and `scontent-cdg2-1`,
the other parts seems stable).

> > @@ -0,0 +1,189 @@
> +# -*- coding: utf-8 -*-
> +
> +# Copyright 2018 Leonardo Taccari
> +#
> +# This program is free software; you can redistribute it and/or modify
> +# it under the terms of the GNU General Public License version 2 as
> +# published by the Free Software Foundation.
> +
> +"""Extract images from Instagram"""
>
> The module docstring should end in the full root URL of the site in question.
> `Instagram` -> `https://www.instagram.com/`
>
> (This is used to auto-generate `docs/supportedsites.rst`. You should run `./scripts/build_supportedsites.py` and see how this turns out.)
>

I've adjusted that.

> > +import hashlib
> +import json
> +from .common import Extractor, Message
> +from .. import text
> +
> +
> +class InstagramExtractor(Extractor):
> +    """Base class for instagram extractors"""
> +    category = "instagram"
> +    directory_fmt = ["{category}", "{username}"]
> +    filename_fmt = "{media_id}.{extension}"
> +    archive_fmt = "{media_id}"
> +    root = "https://www.instagram.com"
> +
> +    def __init__(self, match):
> +        Extractor.__init__(self)
>
> This appears rather pointless. Don't use an `__init__` function if there is nothing to do.
>

I have removed it and adjusted the other `__init__()` to call
`InstagramExtractor.__init__(self)` instead of
`InstagramExtractor.__init__(self, match)`.

> > +                          'window._sharedData = ', ';</script>')[0])
> +
> +    def _extract_postpage(self, url):
> +        page = self.request(url).text
> +        shared_data = self._extract_shared_data(page)
> +        media = shared_data['entry_data']['PostPage'][0]['graphql']['shortcode_media']
> +
> +        return {
> +            'media_id': media['id'],
> +            'shortcode': media['shortcode'],
> +            'typename': media['__typename'],
> +            'display_url': media['display_url'],
> +            'height': int(media['dimensions']['height']),
> +            'width': int(media['dimensions']['width']),
> +            'comments': int(media['edge_media_to_comment']['count']),
> +            'likes': int(media['edge_media_preview_like']['count']),
>
> Use `text.parse_int()` to convert numbers to integer. It doesn't throw an exception if things go wrong.
>

I have replaced all `int()` to `text.parse_int()`.

> > +            end_cursor = timeline['page_info']['end_cursor']
> +            variables = '{{"id":"{}","first":12,"after":"{}"}}'.format(
> +                user_id,
> +                end_cursor,
> +            )
> +            xigis = '{}:{}'.format(rhx_gis, variables)
> +            headers = {
> +                "X-Requested-With": "XMLHttpRequest",
> +                "X-Instagram-GIS": hashlib.md5(xigis.encode()).hexdigest(),
> +            }
> +            url = '{}/graphql/query/?query_hash={}&variables={}'.format(
> +                self.root,
> +                '66eb9403e44cc12e5b5ecda48b667d41',
> +                variables,
> +            )
> +            shared_data = json.loads(self.request(url, headers=headers).text)
>
> The `response` object returned by `self.request()` has a `.json()` method. Use that instead of "manually" loading the JSON result. -> `shared_data = self.request(...).json()`
>

Whoops, I have adjusted it as suggested!

> > +                "media_id": "1923502432034620000",
> +                "shortcode": "Bqxp0VSBgJg",
> +                "typename": "GraphVideo",
> +                "username": "instagram",
> +                "width": int,
> +            }
> +        }),
> +    ]
> +
> +    def __init__(self, match):
> +        InstagramExtractor.__init__(self, match)
> +        self.shortcode = match.group(1)
> +
> +    def instagrams(self):
> +        url = '{}/p/{}/'.format(self.root, self.shortcode)
> +        return [self._extract_postpage(url)]
>
> Not too important, but you can use a tuple instead of a list. (It's a tiny bit more lightweight)
> `[...]` -> `(...,)`
>

I have converted it to a tuple.

> > +        }),
> +    ]
> +
> +    def __init__(self, match):
> +        InstagramExtractor.__init__(self, match)
> +        self.shortcode = match.group(1)
> +
> +    def instagrams(self):
> +        url = '{}/p/{}/'.format(self.root, self.shortcode)
> +        return [self._extract_postpage(url)]
> +
> +
> +class InstagramProfilepageExtractor(InstagramExtractor):
> +    """Extractor for ProfilePage"""
> +    subcategory = "profilepage"
> +    pattern = [r"(?:https?://)?(?:www\.)?instagram\.com/([^/]+)/?$"]
>
> Use a negative lookahead if you don't want to match single `p/`s here.
> `instagram\.com/(?!p/)([^/?&#]+)`
>

TIL about look{ahead,behind}, nice catch, I have adjusted it! :)

I have pushed everything via a separate commit and rerun locally
the tests... I hope there will not be new regressions in the CI,
I'll keep a look to it and adjust accordingly if any!


Thank you very much again Mike for the review and all the feedbacks!
Keep up the good work!
